types/single-spa-react.d.ts

```diff
@@ -1,6 +1,8 @@
 import * as React from "react";
 import * as ReactDOM from "react-dom";
-import { AppProps, LifeCycleFn } from "single-spa";
+import { AppProps, CustomProps, LifeCycleFn } from "single-spa";
+
+export const SingleSpaContext: React.Context<CustomProps & AppProps>;
```

Comment:
I think it would be better to add generics here to let the user specify which custom props they're using. The `CustomProps` type in single-spa lets the props be anything.

Maybe something like this?

```suggestion
export const SingleSpaContext<RootComponentProps>: React.Context<RootComponentProps & AppProps>;
```

Reply:
I agree, that would be better. But I don't know a way to create a generic const. I'm actually not sure which road would be best to take. I can only thing of two options.


**Option 1**
Maybe also export a function that returns `React.useContext(SingleSpaContext)` and give the hook an optional generic type.
```ts
export const SingleSpaContext: React.Context<CustomProps & AppProps>;

export function useSingleSpa<T extends CustomProps = CustomProps>():
  | (T & AppProps)
  | undefined;
```

**Option 2**
Another option could be to use a global interface that can be modified using [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html).
```ts
declare global {
  namespace SingleSpa {
    interface Context extends AppProps {}
  }
}

export const SingleSpaContext: React.Context<SingleSpa.Context>;
```

Then users can modify the type by adding props to the global interface.
```ts
declare global {
  namespace SingleSpa {
    interface Context {
      hello: string;
      world: number;
    }
  }
}
```

Reply:
Ah good point, I didn't think that through about adding a generic to props. I don't think Option 1 makes sense to implement, but option 2 might be alright. However, the namespace wouldn't be `SingleSpa` but `SingleSpaReact`.

All of this said, though, I think it would be fine to implement neither of the changes and just leave it without the generics. The user of SingleSpaContext should be able to specify the type of the context as a generic argument to `useContext<CustomType>(SingleSpaContext)`. And I think that's probably the best way.

Could you add tests for the types, as described above? After that I'll approve and merge.

Reply:
You will have to use `useContext(SingleSpaContext) as CustomType`. Using the generic option requires SingleSpaContext to be a `React.Context<CustomType>`.

I've added some tests!

Reply:
Yeah good point, type casting will be needed to get it to be fully strict. I think that's good enough for now, though - we can improve later if someone finds a good way to do so.